symengine/lib/symengine_wrapper.pyx

```diff
@@ -2078,22 +2057,137 @@ class DenseMatrixIter(object):
+        for j in range(self.rows):
+            for k in range(self.cols - 1):
+                if k >= i:
+                    result[j, k] = self[j, k + 1]
+                else:
+                    result[j, k] = self[j, k]
+        return result
+
+    def col_op(self, j, f):
+        raise NotImplementedError
+
+    def col_swap(self, i, j):
+        for k in range(0, self.rows):
+            self[k, i], self[k, j] = self[k, j], self[k, i]
+
+    def copyin_list(self, key, value):
+        raise NotImplementedError
+
+    def copyin_matrix(self, key, value):
+        rlo, rhi, clo, chi = self.key2bounds(key)
+        shape = value.shape
+        dr, dc = rhi - rlo, chi - clo
+        if shape != (dr, dc):
+            raise ShapeError("The Matrix `value` doesn't have the same"\
+                             "dimensions as the in sub-Matrix given by `key`.")
+
+        for i in range(value.rows):
+            for j in range(value.cols):
+                self[i + rlo, j + clo] = value[i, j]
+
+    def fill(self, value):
+        for i in range(self.rows):
+            for j in range(self.cols):
+                self[i, j] = value
+
+    def row_del(self, i):
+        if i < -self.rows or i >= self.rows:
+            raise IndexError("Index out of range: 'i = %s', valid -%s <= i"
+                             " < %s" % (i, self.rows, self.rows))
+        if i < 0:
+            i += self.rows
+        cdef DenseMatrixBase result = zeros(self.rows - 1, self.cols)
+        for j in range(self.rows - 1):
+            for k in range(self.cols):
+                if j >= i:
+                    result[j, k] = self[j + 1, k]
+                else:
+                    result[j, k] = self[j, k]
+        return result
+
+    def row_op(self, i, f):
+        raise NotImplementedError
+
+    def row_swap(self, i, j):
+        for k in range(0, self.cols):
+            self[i, k], self[j, k] = self[j, k], self[i, k]
+
+    def simplify(self, *args):
+        raise NotImplementedError
+
+    def zip_row_op(self, i, k, f):
+        raise NotImplementedError
+
+Matrix = DenseMatrix = MutableDenseMatrix
+
+cdef class ImmutableDenseMatrix(DenseMatrixBase):
+
+    def _entry(self, i, j):
+        return DenseMatrixBase.__getitem__(self, (i, j))
+
+    def __setitem__(self, key, value):
+        raise TypeError("Cannot set values of {}".format(self.__class__))
+
+    def set(self, i, j, e):
+        raise TypeError("Cannot set values of {}".format(self.__class__))
+
+    def _set(self, i, j, e):
+        raise TypeError("Cannot set values of {}".format(self.__class__))
+
+ImmutableMatrix = ImmutableDenseMatrix
+
+cdef matrix_to_vec(DenseMatrixBase d, symengine.vec_basic& v):
     cdef Basic e_
     for i in range(d.nrows()):
         for j in range(d.ncols()):
             e_ = d._get(i, j)
             v.push_back(e_.thisptr)
 
 def eye(n):
-    d = DenseMatrix(n, n)
+    d = DenseMatrixBase(n, n)
```

Comment:
Idea is that `DenseMatrixBase` should not be initialized directly at all. Only one of `MutableDenseMatrix` or `ImmutableDenseMatrix` should be used